File: packages/emitter/testcases/classes/abstract/AbstractClasses.cs
namespace TestCases.classes.abstract
{
    public class Shape
    {
        public double getArea();

        public double getPerimeter();

        public string describe()
            {
            return $"Area: {this.getArea()}, Perimeter: {this.getPerimeter()}";
            }
    }
    public class Rectangle : Shape
    {
        public double width;

        public double height;

        public Rectangle(double width, double height) : base()
            {
            this.width = width;
            this.height = height;
            }

        public override double getArea()
            {
            return this.width * this.height;
            }

        public override double getPerimeter()
            {
            return 2.0 * (this.width + this.height);
            }
    }
    public class Circle : Shape
    {
        public double radius;

        public Circle(double radius) : base()
            {
            this.radius = radius;
            }

        public override double getArea()
            {
            return global::Tsonic.JSRuntime.Math.PI * this.radius * this.radius;
            }

        public override double getPerimeter()
            {
            return 2.0 * global::Tsonic.JSRuntime.Math.PI * this.radius;
            }
    }

            public static class AbstractClasses
            {
                public static double calculateTotalArea(global::System.Collections.Generic.List<Shape> shapes)
                    {
                    return global::Tsonic.JSRuntime.Array.reduce(shapes, (double total, Shape shape) => total + shape.getArea(), 0.0);
                    }
            }
}

File: packages/emitter/testcases/js-only/classes/abstract/AbstractClasses.ts
// Abstract base class
export abstract class Shape {
  abstract getArea(): number;
  abstract getPerimeter(): number;

  // Concrete method in abstract class
  describe(): string {
    return `Area: ${this.getArea()}, Perimeter: ${this.getPerimeter()}`;
  }
}

// Concrete implementation
export class Rectangle extends Shape {
  width: number;
  height: number;

  constructor(width: number, height: number) {
    super();
    this.width = width;
    this.height = height;
  }

  getArea(): number {
    return this.width * this.height;
  }

  getPerimeter(): number {
    return 2 * (this.width + this.height);
  }
}

// Another concrete implementation
export class Circle extends Shape {
  radius: number;

  constructor(radius: number) {
    super();
    this.radius = radius;
  }

  getArea(): number {
    return Math.PI * this.radius * this.radius;
  }

  getPerimeter(): number {
    return 2 * Math.PI * this.radius;
  }
}

// Function using abstract class
export function calculateTotalArea(shapes: Shape[]): number {
  return shapes.reduce(
    (total: number, shape: Shape): number => total + shape.getArea(),
    0
  );
}
